jgrapht-core/src/main/java/org/jgrapht/generate/BarabasiAlbertGraphGenerator.java

```diff
@@ -0,0 +1,203 @@
+     * @throws IllegalArgumentException in case of invalid parameters
+     */
+    public BarabasiAlbertGraphGenerator(int m0, int m, int n, long seed)
+    {
+        this(m0, m, n, new Random(seed));
+    }
+
+    /**
+     * Constructor
+     * 
+     * @param m0 number of initial nodes
+     * @param m number of edges of each new node added during the network growth
+     * @param n final number of nodes
+     * @param rng the random number generator to use
+     * @throws IllegalArgumentException in case of invalid parameters
+     */
+    public BarabasiAlbertGraphGenerator(int m0, int m, int n, Random rng)
+    {
+        if (m0 < 1) {
+            throw new IllegalArgumentException("invalid initial nodes (" + m0 + " < 1)");
+        }
+        this.m0 = m0;
+        if (m <= 0) {
+            throw new IllegalArgumentException("invalid edges per node (" + m + " <= 0");
+        }
+        if (m > m0) {
+            throw new IllegalArgumentException("invalid edges per node (" + m + " > " + m0 + ")");
+        }
+        this.m = m;
+        if (n < m0) {
+            throw new IllegalArgumentException(
+                "total number of nodes must be at least equal to the initial set");
+        }
+        this.n = n;
+        this.rng = Objects.requireNonNull(rng, "Random number generator cannot be null");
+    }
+
+    /**
+     * Generates an instance.
+     * 
+     * @param target the target graph
+     * @param vertexFactory the vertex factory
+     * @param resultMap not used by this generator, can be null
+     */
+    @Override
+    public void generateGraph(
+        Graph<V, E> target, VertexFactory<V> vertexFactory, Map<String, V> resultMap)
+    {
+        /*
+         * Ensure directed or undirected.
+         */
+        boolean isDirected;
+        if (target instanceof DirectedGraph<?, ?>) {
+            isDirected = true;
+        } else if (target instanceof UndirectedGraph<?, ?>) {
+            isDirected = false;
+        } else {
+            throw new IllegalArgumentException("Graph must be either directed or undirected");
+        }
+
+        /*
+         * Create complete graph with m0 nodes
+         */
+        List<V> nodes = new ArrayList<>(n * m);
+        for (int i = 0; i < m0; i++) {
+            V v = vertexFactory.createVertex();
+            if (!target.addVertex(v)) {
+                throw new IllegalArgumentException("Invalid vertex factory");
+            }
+            nodes.add(v);
+        }
+        for (int i = 0; i < m0; i++) {
+            for (int j = i + 1; j < m0; j++) {
+                V v = nodes.get(i);
+                V u = nodes.get(j);
+                target.addEdge(v, u);
+                if (isDirected) {
+                    target.addEdge(u, v);
+                }
+            }
+        }
+
+        /*
+         * Augment node list to have node multiplicity equal to m0-1.
+         */
+        for (int i = 0; i < m0 - 2; i++) {
```

Comment:
It is! The 'n*m' is the maximum that it will get. The nodes list contains each node multiple times (once per edge target) in order to sample based on the degrees.